TechnicalDocuments/Proposals/IdentifiersForUIAutomation.md

```diff
@@ -0,0 +1,203 @@
+
+```swift
+extension DesignLibrary {
+    public enum AutomationName {}
+}
+```
+
+For example, given the conformance of these two atomic components:
+
+```swift
+struct UI.Label {
+    // ... other properties
+    let automationName: String
+
+    init(/* ..., */automationName: String = "label") {
+        self.automationName = automationName
+    }
+}
+
+struct UI.Button {
+    // ... other properties
+    let automationName: String
+
+    init(/* ..., */automationName: String = "button") {
+        self.automationName = automationName
+    }
+}
+
+extension UI.Label: ViewLifecycleAware {
+    func updateAccessibilityIdentifiers(usingPrefix prefix: String) {
+        return prefix + automationName
+    }
+}
+
+extension UI.Button: ViewLifecycleAware {
+    func updateAccessibilityIdentifiers(usingPrefix prefix: String) {
+        return prefix + automationName
+    }
+}
+```
+
+and this particular composite component called `prettyCard`:
+
+```swift
+extension DesignLibrary.ComponentsLibrary {
+    public func prettyCard() -> AnyRenderable {
+        return [
+            UI.Label(/* ... */, automationName: String(AutomationName.prettyCardNotice)),
+            UI.Button(/* ... */, automationName: String(AutomationName.prettyCardSubmit))
+        ].stack(axis: .vertical)
+    }
+}
+
+extension DesignLibrary.AutomationName {
+    public enum PrettyCard: String {
+        case notice
+        case submit
+    }
+}
+```
+
+When we render `prettyCard`:
+
+```
+render(
+    Box(
+        sections: [
+            Section(id: .activity, items: [
+                Node(id: .card, component:
+                    prettyCard()
+                )
+            ])
+        ]
+    )
+)
+```
+
+It should result in two accessibility identifiers:
+
+* `activity/card/notice` applied to the label; and
+* `activity/card/submit` applied to the button.
+
+Notice that automation names for the relevant subcomponents of `prettyCard` has been published in the `DesignLibrary.AutomationName` umbrella. This facilitates the next suggestion of this proposal...
+
+### Working with UI automation tests
+
+Now that we have a systematic approach of defining and publishing information about accessibility identifiers, we can use this to reliably generate accessibility identifiers in UI automation tests!
+
+The proposal suggests that the UI automation test target should link with all UI frameworks. This allows the automation code to have access to `Renderer`s and any published `AutomationName` enums, and can therefore construct accessibility identifiers. Changes made to any section ID, item ID or automation name enums would either be automatically picked up by the automation code, or lead to compile-time breakage depending on the circusmtances.
+
+`BaseScreen` shall be extended to provide a conveinence:
+```swift
+extension BaseScreen {
+    subscript<Renderer: BoxRenderer>(
+        _ renderer: Renderer.Type,
```

Comment:
why `renderer` parameter is needed? `Renderer.SectionID` is not enough for compiler?

Reply:
Swift compiler does not work out the type backwards AFAIK. Even if you write:

```swift
self[FeedRenderer.SectionID.header, FeedRenderer.ItemID.headerTitle]
```

From the type system's perspective, the ID types can be used by multiple types for the same associated type location.

Also it is less verbose with the proposed subscript, which enables contextual lookup at the 2nd and 3rd arguments. 
```swift
self[FeedRenderer.self, .header, .headerTitle]
```

Reply:
> it is less verbose/noisy

it's still more verbose and noisy than `self[.header, .headerTitle]`. I still feel there is a room for improvement here by associating Screen object with Renderer type (which is 1-to-1 relationship anyway in most cases, if it's not - it should be) rather than requiring to specify it in each subscript, and having a subscript like `screen[.header, .headerTitle]` where `screen` would be a helper wrapper generic over renderer type. Roughly this:

```
class FeedScreen {
  let screen = ScreenLookup<FeedRenderer>(app)
}

class ScreenLookup<Renderer: BoxRenderer> {
  subscript(_ sectionId: Renderer.SectionID, _ nodeId: Renderer.NodeId) -> XCUIElement {
    return app[...]
  }
}
```

Reply:
We can't go simpler than that unless we have that type information on the screen object or some kind of proxy object, like your snippet.